pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,433 @@
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCPodRouteController {
+	controller := &MCPodRouteController{
+		k8sClient:       k8sClient,
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		podQueue:        workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForPod"),
+		gwQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteControllerForGateway"),
+		gwInformer:      gwInformer,
+		gwLister:        mclisters.NewGatewayLister(gwInformer.GetIndexer()),
+		namespace:       namespace,
+		podCache:        make(map[string]*podNodeIPs),
+		ipToPods:        make(map[string]sets.String),
+		podWorkerStopCh: make(chan struct{}),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *MCPodRouteController) createPodInformer() {
+	listOptions := func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermNotEqualSelector("spec.nodeName", c.nodeConfig.Name).String()
+	}
+	c.podInformer = coreinformers.NewFilteredPodInformer(
+		c.k8sClient,
+		metav1.NamespaceAll,
+		0,
+		nil,
+		listOptions,
+	)
+	c.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				c.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				c.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				c.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	c.podLister = corelisters.NewPodLister(c.podInformer.GetIndexer())
+}
+
+func (c *MCPodRouteController) enqueueGateway(obj interface{}) {
+	_, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueueGateway can't process event", "obj", obj)
+		return
+	}
+	c.gwQueue.Add(dummyKey)
+}
+
+// validatePod validates the Pod info to determine if it's a valid Pod which should
+// have	Openflow flows installed for cross-cluster traffic.
+func (c *MCPodRouteController) validatePod(pod *corev1.Pod) bool {
+	// pod.Status.PodIP and HostIP should be an IPv4 address for IPv4 and dual stack clusters.
+	// IPv6 only cluster is not supported by Antrea Multi-cluster yet.
+	podIP := pod.Status.PodIP
+	podNodeIP := pod.Status.HostIP
+
+	if pod.Spec.NodeName == c.nodeConfig.Name {
+		klog.V(2).InfoS("Skip Pod on the current Node")
+		return false
+	}
+
+	if podNodeIP == "" || pod.Spec.HostNetwork {
```

Comment:
I feel this is not safe, as the Pod may be updated to be hostNetwork and then the previously installed flows will not be removed. We should also check if the Pod has been cached and if the Pod IP or Node IP has changed; if the Pod is not cached, then check Node IP and HostNetwork.

Reply:
Pod's hostNetwork field can't be updated when Pod is already created. If the Pod is a part of a deployment, it will be deleted first, then created.

Reply:
Pod delete and recreate can be merged to one event AFAIK. And here quite unnecessary to make such assumption on K8s behaviors, when you already have a Pod cache and can just check the cache to know the status.

Reply:
Done, add a step to check cache first.